### ISSUE TITLE
Request: aws_regions (plural) data source which can be filtered by product.

### ISSUE THREAD

**Issue by apowers** [terraform-providers/terraform-provider-aws#3763]
Feature Request:
aws_regions (plural) data source which can be filtered by product.

Example of desired feature to return a list of regions that support EC2.
```
data "aws_regions" "ec2" {}
```

**Comment by bflad** [comment]
Sounds like an interesting use case! Are you able to explain how you were planning on using this within Terraform?

This seems like it might be possible using the `endpoints` data vendored with the AWS SDK (caveat: only up to date with the SDK version baked into the provider at time of release), although it would probably require an implementation like the below to keep it generic for supporting other services:

```hcl
# Not implemented - may have different interface in the future
data "aws_regions" "ec2" {
  service_name = "ec2"
}

output "ec2_regions" {
  value = "${join(",", data.aws_regions.ec2.regions)}"
}
```

**Comment by apowers** [comment]
Thank you for your quick reply.

The security team wants to deploy their auditing application and ensure that it is running in every region. When new regions are available, they want it to be deployed in those new regions automatically.

**Comment by bflad** [comment]
There is a caveat though, I do not believe its possible to automatically configure `provider`s in this fashion as `provider` does not support `count` as if Terraform 0.11.3. See also: https://github.com/hashicorp/terraform/issues/16967

**Comment by bflad** [comment]
Hi folks 👋  A new [`aws_regions` data source](https://www.terraform.io/docs/providers/aws/d/regions.html) was previously released in [version 2.58.0 of the Terraform AWS Provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2580-april-16-2020). It uses the [EC2 `DescribeRegions` API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeRegions.html) to return a set of available regions, potentially filtered by their opt-in state. To further track Terraform support for declaring multiple providers with  `count`/`for_each`-like syntax, please follow https://github.com/hashicorp/terraform/issues/16967.
